File: dotfiles/.config/sway/status.py
#!/usr/bin/env python3
import json, time, sys, subprocess, os

def get_status():
    items = []

    # Network
    try:
        ip = subprocess.check_output(["ip", "-4", "-o", "addr", "show", "scope", "global"], text=True).strip()
        if ip:
            iface, addr = ip.split()[1], ip.split()[3].split("/")[0]
            items.append({"full_text": f"{iface}: {addr}", "color": "#0055aa"})
        else:
            items.append({"full_text": "no network", "color": "#990000"})
    except:
        items.append({"full_text": "no network", "color": "#990000"})

    # Disk
    st = os.statvfs("/")
    free_gb = (st.f_bavail * st.f_frsize) / (1024**3)
    items.append({"full_text": f"disk: {free_gb:.1f}G"})

    # Load
    load = os.getloadavg()[0]
    items.append({"full_text": f"load: {load:.2f}"})

    # Memory
    with open("/proc/meminfo") as f:
        mem = {}
        for line in f:
            k, v = line.split(":")[0], int(line.split(":")[1].strip().split()[0])
            mem[k] = v
    used = (mem["MemTotal"] - mem["MemAvailable"]) / 1024
    total = mem["MemTotal"] / 1024
    items.append({"full_text": f"mem: {used:.0f}/{total:.0f}M"})

    # Time
    items.append({"full_text": time.strftime("%Y-%m-%d %H:%M:%S")})

    # Right margin
    items.append({"full_text": " "})

    return items

print(json.dumps({"version": 1}))
print("[")
print("[]")
sys.stdout.flush()

while True:
    print("," + json.dumps(get_status()))
    sys.stdout.flush()
    time.sleep(5)

File: dotfiles/.config/sway/status.c
#define _DEFAULT_SOURCE
#include <stdio.h>
#include <string.h>
#include <time.h>
#include <unistd.h>
#include <sys/statvfs.h>
#include <sys/sysinfo.h>

int main() {
    printf("{\"version\":1}\n[\n[]\n");
    fflush(stdout);
    while (1) {
        char ip[96] = "no network";
        FILE *f = popen("ip -4 -o addr show scope global 2>/dev/null", "r");
        if (f) {
            char buf[256], iface[32], addr[64];
            if (fgets(buf, sizeof(buf), f))
                if (sscanf(buf, "%*d: %31s inet %63[^/]", iface, addr) == 2)
                    snprintf(ip, sizeof(ip), "%s: %s", iface, addr);
            pclose(f);
        }

        struct statvfs sv;
        statvfs("/", &sv);
        double disk = (double)sv.f_bavail * sv.f_frsize / (1024*1024*1024);

        double load0 = 0;
        FILE *lf = fopen("/proc/loadavg", "r");
        if (lf) { fscanf(lf, "%lf", &load0); fclose(lf); }

        struct sysinfo si;
        sysinfo(&si);
        long used = (si.totalram - si.freeram - si.bufferram - si.sharedram) / 1024 / 1024;
        long total = si.totalram / 1024 / 1024;

        char timebuf[32];
        time_t t = time(NULL);
        strftime(timebuf, sizeof(timebuf), "%Y-%m-%d %H:%M:%S", localtime(&t));

        printf(",["
            "{\"full_text\":\"%s\",\"color\":\"#0055aa\"},"
            "{\"full_text\":\"disk: %.1fG\"},"
            "{\"full_text\":\"load: %.2f\"},"
            "{\"full_text\":\"mem: %ld/%ldM\"},"
            "{\"full_text\":\"%s\"},"
            "{\"full_text\":\" \"}"
            "]\n", ip, disk, load0, used, total, timebuf);
        fflush(stdout);
        sleep(5);
    }
}
